### ISSUE TITLE
[Does xm.master_print() print the performance of the model on 1/num_cores of the validation data?]

### ISSUE THREAD

**Issue by MouathAb** [pytorch/xla#2856]
XLA backend : TPU (google colab)
 torch_xla version: 1.7
 model: Vision Transformer 

The model accuracy after restoring on TPU is changing.
I think that the printed accuracy with xm.master_print [during train/valid] is the accuracy on only 1/num_cores (num_cores=8 in my case) of the validation dataset; which may be different when evaluating the saved model on the whole validation set.

This is a part of the code for train/valid:
`    for epoch in range(1, epochs + 1):
        gc.collect()
        para_train_loader = pl.ParallelLoader(train_loader, [device])

        xm.master_print(f"{'='*50}")
        xm.master_print(f"EPOCH {epoch} - TRAINING...")
        
        train_loss, train_acc = model.train_one_epoch(para_train_loader.per_device_loader(device), criterion, optimizer, device,scheduler)
        xm.master_print(
            f"\n\t[TRAIN] EPOCH {epoch} - LOSS: {train_loss}, ACCURACY: {train_acc}\n"
        )
        train_losses.append(train_loss)
        train_accs.append(train_acc)
        gc.collect()
        
        # evaluate the model on the validation data
        if valid_loader is not None:
            gc.collect()
            para_valid_loader = pl.ParallelLoader(valid_loader, [device])
            xm.master_print(f"EPOCH {epoch} - VALIDATING...")
            valid_loss, valid_acc = model.validate_one_epoch(
                para_valid_loader.per_device_loader(device), criterion, device)
            xm.master_print(f"\t[VALID] LOSS: {valid_loss}, ACCURACY: {valid_acc}\n")
            valid_losses.append(valid_loss)
            valid_accs.append(valid_acc)
            gc.collect()`

And this is the inference part where I re-test the saved model on the validation dataset  :

`valid_dataset = Dataset(valid_df, transforms=transforms_valid)
valid_sampler = torch.utils.data.distributed.DistributedSampler(valid_dataset,num_replicas=xm.xrt_world_size(),rank=xm.get_ordinal(),shuffle=False)
valid_loader = torch.utils.data.DataLoader(dataset=valid_dataset,batch_size=len(valid_df),sampler=valid_sampler,drop_last=True,num_workers=8)

model = TRAIN(n_classes=num_classes, pretrained=True) # create the model
device = xm.xla_device()
model.to(device)
model.load_state_dict(torch.load(Path_to_saved_modell))
model.eval()
criterion = nn.CrossEntropyLoss()

data, target = next(iter(valid_loader))
data = data.to(device, dtype=torch.float32)
target = target.to(device, dtype=torch.int64)

with torch.no_grad():

    # compute predicted outputs by passing inputs to the model
    output = model(data)
    pred = output.argmax(dim=1)

    # calculate the batch loss
    loss = criterion(output, target)

    # Calculate Accuracy
    accuracy = (pred == target).float().mean()

print(accuracy)
`

PS : I tried one  thing that works but .., during valid/train process I validate on a vf dataFrame which is the valid_df duplicated 8 times as the number of cores and I set the batch size as the length of valid_df: 

`vf = pd.concat([valid_df,valid_df,valid_df,valid_df,valid_df,valid_df,valid_df,valid_df],ignore_index=True)

valid_dataset = Dataset(vf, transforms=transforms_valid)

valid_sampler = torch.utils.data.distributed.DistributedSampler(
            valid_dataset,
            num_replicas=xm.xrt_world_size(),
            rank=xm.get_ordinal(),
            shuffle=False,
        )

valid_loader = torch.utils.data.DataLoader(
            dataset=valid_dataset,
            batch_size = len(valid_df),
            sampler=valid_sampler,
            drop_last=True,
            num_workers=8,
        )
` 
in that way, I could have the same performance.

**Comment by taylanbil** [comment]
`xm.master_print` will print whatever data the main process has that the code is printing. If this is only part of the total validation data (standard practice), then `xm.master_print` will print that.

In the code above, if I understand correctly, `valid_sampler` shards the total validation data, so the main process will have 1/num_cores of the total validation data. Thus, `xm.master_print` will print the accuracy belonging to this slice of the validation dataset.

If you wanted to see accuracies from all cores, you could use `print` instead. Or, you could reduce the accuracies from all cores as [here](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L228-L229), and use `xm.master_print` to print the average accuracy from all cores.
